ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java

```diff
@@ -123,9 +128,19 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
         final DataStreamReply reply = (DataStreamReply) msg;
         LOG.debug("{}: read {}", this, reply);
         clientInvocationId = ClientInvocationId.valueOf(reply.getClientId(), reply.getStreamId());
-        Optional.ofNullable(replies.get(clientInvocationId))
-            .map(Queue::poll)
-            .ifPresent(f -> f.complete(reply));
+        final Queue<CompletableFuture<DataStreamReply>> queue = replies.get(clientInvocationId);
+        if (queue != null) {
+          final CompletableFuture<DataStreamReply> f = queue.poll();
+          if (f != null) {
+            f.complete(reply);
+            if (queue.isEmpty()) {
+              timeoutScheduler.onTimeout(emptyQueueGracePeriod,
+                  // remove the queue if the same queue has been empty for the entire grace period.
+                  () -> replies.computeIfPresent(clientInvocationId, (id, q) -> q == queue && q.isEmpty()? null: q),
```

Comment:
I just found that the queue could have changed from empty to non-empty and then from non-empty to empty in between.  We need an "empty id" to detect the change; see https://issues.apache.org/jira/secure/attachment/13036796/542_review2.patch